liferay-headless/liferay-portal
# LPS-191058 Add filter support in API Builder Endpoints to Collections and related elements of the main object definition

Related issue: https://liferay.atlassian.net/browse/LPS-191058

NOTE: This PR has been created on top of brianchandotcom/master branch, so it is only necessary to review the last set of commits tagged with LPS-191058

---

The purposes of this PR are the following ones:

- Support filtering (using the `filter` query parameter in the API Endpoints of an API Application) over the fields that are configured in the API Schema as fields **of a related object definition** to the main object definition associated to the API Schema. That is, supporting the `ComplexEntityField` of the main object definition's EntityModel when there is a parameter in the API Schema properly configured.
- Support filtering over collections in API Endpoints of an API Application. That is, for example, filtering over a Multi select picklist field.
- Including integration tests for all the possible operands, not only filtering by using properties of the main object definition associated to the API Schema but also over the related elements of the main object definition.

---

How to test it:

- Rebase this branch over the latest master
- Perform an `ant all` to ensure all the dependencies have been deployed properly
- Execute the integration tests of the class `HeadlessBuilderResourceTest` of the module `modules/apps/headless/headless-builder/headless-builder-test` with the command `gw tI --tests HeadlessBuilderResourceTest`
